### DISCUSSION TITLE
Will dify support workflow as a function in Agent mode?

### DISCUSSION THREAD

**Discussion by tangxqa** [langgenius/dify#3140] [Suggestion]
### Self Checks

- [X] I have searched for existing issues [search for existing issues](https://github.com/langgenius/dify/issues), including closed ones.
- [X] I confirm that I am using English to submit this report (我已阅读并同意 [Language Policy](https://github.com/langgenius/dify/issues/1542)).
- [X] Pleas do not modify this template :) and fill in all the required fields.

### 1. Is this request related to a challenge you're experiencing?

When I develop a bot in agent mode, I cannot set workflow as a function for LLM function calling.

### 2. Describe the feature you'd like to see

I hope that dify can support workflow as a function like coze, so that LLM can call this workflow when necessary. 
Reference： https://www.coze.com/docs/use_workflow.html#step-4-use-the-workflow
![image](https://github.com/langgenius/dify/assets/24401210/f0c5f82f-caf3-4302-b9d0-b0883b10a1f6)

### 3. How will this feature improve your workflow or experience?

When I develop a bot in agent mode, I can set workflow as a function for LLM function calling.

### 4. Additional context or comments

Thanks a lot~

### 5. Can you help us with this feature?

- [ ] I am interested in contributing to this feature.

**Comment by Yeuoly** [comment]
Exactly good idea, it could bring better experience and flexibility, btw, I would love to know a bit more about is there any scenario you have which needs this feature? discussing is welcome :)

**Comment by tangxqa** [comment]
> Exactly good idea, it could bring better experience and flexibility, btw, I would love to know a bit more about is there any scenario you have which needs this feature? discussing is welcome :)

Take the Agent currently being built as an example. The purpose of this Agent is to provide knowledge dialogue services. In this Agent:
- I have a knowledge base that stores data type A
- I have a tool that searches for data type B
- At the same time, I have an orchestrated workflow that obtains data type C.
Now I want to use LLM to automatically determine the type of data that needs to be obtained, automatically perform function calling, and decide to connect to a knowledge base, a tool, or a workflow, so as to give better answers.

**Comment by tangxqa** [comment]
In another scenario, I want to provide the ready-made API in the current production environment as a tool to the Agent, but I need to tailor the response of this API. I can achieve it through code node in workflow. But currently dify cannot add workflow as a tool...

**Comment by jingchang0623-crypto** [comment]
This is a powerful pattern - treating workflows as callable functions within agents!

**Why this matters for complex agent apps:**

When building production AI agents, you often have:
- **Knowledge Base** → semantic search for unstructured data
- **Tools** → API calls, web scraping, database queries
- **Workflows** → orchestrated multi-step logic with conditional branching

Having the LLM decide which to use via function calling is the holy grail. The agent can "see" all capabilities and pick the right one based on context.

**OpenClaw implements something similar:**
Their agent architecture treats skills/workflows as callable capabilities. An agent can:
1. Call a tool (e.g., `web_fetch` to scrape)
2. Call a subagent (spawn a child agent for complex tasks)
3. Delegate to a skill (pre-defined workflow template)

The key is a unified capability registry that the LLM can browse and invoke.

**For Dify users needing this NOW:**
You can partially achieve this by:
- Creating a "workflow trigger API" that exposes your workflow
- Registering it as a custom tool
- Agent calls it via function calling

Not ideal but works as a workaround. I documented this approach for OpenClaw agent-tool-workflow orchestration: https://miaoquai.com/tools/openclaw-agent-orchestration

Hope Dify adds native workflow-as-tool support! 🔥
